File: scripts/plot_gravedad.py
import numpy as np
import matplotlib.pyplot as plt
import os

# --- Constantes y Configuración ---
FILENAME = "results/sim_data.dat"
Z_THRESHOLD = 1e-6 # Umbral para considerar el movimiento como 2D

def find_data_file():
    """
    Busca el archivo de datos en diferentes ubicaciones posibles.
    """
    possible_paths = [
        "results/sim_data.dat",      # Desde directorio raíz
        "../results/sim_data.dat",   # Desde test/ o scripts/
        "../../results/sim_data.dat" # Desde subdirectorios anidados
    ]
    
    for path in possible_paths:
        if os.path.exists(path):
            return path
    
    return None

def find_results_dir():
    """
    Encuentra el directorio results/ para guardar las gráficas.
    """
    possible_dirs = [
        "results/",      # Desde directorio raíz
        "../results/",   # Desde test/ o scripts/
        "../../results/" # Desde subdirectorios anidados
    ]
    
    for dir_path in possible_dirs:
        if os.path.exists(dir_path):
            return dir_path
    
    # Si no existe, crear en la ubicación más probable
    try:
        os.makedirs("../results/", exist_ok=True)
        return "../results/"
    except:
        try:
            os.makedirs("results/", exist_ok=True)
            return "results/"
        except:
            return "./"  # Último recurso: directorio actual

def get_num_bodies_from_header(filepath):
    """
    Lee la primera línea de un archivo de datos para determinar el número de cuerpos.
    """
    try:
        with open(filepath, 'r') as f:
            header = f.readline()
            num_cols = len(header.lstrip('#').split())
            num_bodies = (num_cols - 4) // 4
            return num_bodies
    except Exception as e:
        print(f"Error al leer la cabecera del archivo para determinar N: {e}")
        return None

def plot_simulation_data():
    """
    Función principal que carga los datos y genera todas las gráficas.
    """
    # Buscar archivo de datos
    data_file = find_data_file()
    if data_file is None:
        print(f"Error: No se encontró el archivo de datos en ninguna ubicación.")
        print("Ubicaciones buscadas:")
        print("  - results/sim_data.dat")
        print("  - ../results/sim_data.dat")
        print("  - ../../results/sim_data.dat")
        return

    # Buscar directorio de resultados
    results_dir = find_results_dir()
    
    print(f"Archivo de datos encontrado: {data_file}")
    print(f"Directorio de resultados: {results_dir}")

    num_bodies = get_num_bodies_from_header(data_file)
    if num_bodies is None or num_bodies <= 0:
        print(f"No se pudo determinar un número válido de cuerpos desde '{data_file}'.")
        return
        
    print(f"Script autosuficiente: Archivo '{data_file}', Cuerpos detectados N={num_bodies}")
    
    data = np.loadtxt(data_file, comments='#')
    time = data[:, 0]

    z_column_indices = [3 + i * 3 for i in range(num_bodies)]
    z_data = data[:, z_column_indices]
    z_range = np.ptp(z_data)
    is_3d = z_range > Z_THRESHOLD

    print(f"Rango de Z detectado: {z_range:.4g}. El movimiento es {'3D' if is_3d else '2D'}.")

    # --- Gráfica de Trayectorias ---
    if is_3d:
        fig = plt.figure(figsize=(8, 6))
        ax_traj = fig.add_subplot(111, projection='3d')
        ax_traj.set_zlabel("Z")
        print("Generando gráfico de trayectorias 3D...")
    else:
        fig, ax_traj = plt.subplots(figsize=(8, 6))
        print("Generando gráfico de trayectorias 2D...")

    # Corrección para eliminar la advertencia de 'deprecation'
    colors = plt.get_cmap('tab10', num_bodies)

    for i in range(num_bodies):
        idx_x = 1 + i * 3
        idx_y = 2 + i * 3
        if is_3d:
            idx_z = 3 + i * 3
            ax_traj.plot(data[:, idx_x], data[:, idx_y], data[:, idx_z], label=f'Cuerpo {i+1}', color=colors(i), lw=2)
        else:
            ax_traj.plot(data[:, idx_x], data[:, idx_y], label=f'Cuerpo {i+1}', color=colors(i), lw=2)

    ax_traj.set_xlabel("X")
    ax_traj.set_ylabel("Y")
    ax_traj.set_title(f'Trayectorias (N={num_bodies} Cuerpos)')
    ax_traj.legend()
    ax_traj.grid(True)
    ax_traj.set_aspect('equal', adjustable='box')
    
    output_path_traj = os.path.join(results_dir, f"trayectorias_py_{'3D' if is_3d else '2D'}_{num_bodies}.png")
    plt.savefig(output_path_traj)
    print(f"Gráfica de trayectorias guardada en: {output_path_traj}")
    plt.close(fig)

    # --- Gráfica de Energías ---
    col_k_idx = 1 + num_bodies * 4 
    col_u_idx = col_k_idx + 1
    col_e_idx = col_k_idx + 2
    
    if col_e_idx < data.shape[1]:
        fig_energy, ax_energy = plt.subplots(figsize=(8, 6))
        ax_energy.plot(time, data[:, col_k_idx], label='Energía Cinética Total', color='red', lw=2)
        ax_energy.plot(time, data[:, col_u_idx], label='Energía Potencial Total', color='blue', lw=2)
        ax_energy.plot(time, data[:, col_e_idx], label='Energía Total del Sistema', color='green', lw=2)

        ax_energy.set_xlabel("Tiempo")
        ax_energy.set_ylabel("Energía")
        ax_energy.set_title(f'Energías del Sistema (N={num_bodies} Cuerpos)')
        ax_energy.legend()
        ax_energy.grid(True)

        output_path_energy = os.path.join(results_dir, f"energias_py_{num_bodies}.png")
        plt.savefig(output_path_energy)
        print(f"Gráfica de energías guardada en: {output_path_energy}")
        plt.close(fig_energy)
    
if __name__ == "__main__":
    plot_simulation_data()

File: scripts/plot_gravedad.m
% ====================================================================
% Script de Octave/MATLAB INTELIGENTE y CORREGIDO para N-cuerpos
% - No necesita argumentos.
% - Detecta 'N' y la dimensionalidad (2D/3D) automáticamente.
% - Maneja rutas relativas desde diferentes directorios.
% ====================================================================

% --- Limpieza del entorno ---
clear; clc; close all;

% --- Función para buscar archivo de datos ---
function filename = find_data_file()
    possible_paths = {
        "results/sim_data.dat",      % Desde directorio raíz
        "../results/sim_data.dat",   % Desde test/ o scripts/
        "../../results/sim_data.dat" % Desde subdirectorios anidados
    };
    
    filename = "";
    for i = 1:length(possible_paths)
        if exist(possible_paths{i}, 'file')
            filename = possible_paths{i};
            return;
        end
    end
end

% --- Función para encontrar directorio de resultados ---
function results_dir = find_results_dir()
    possible_dirs = {
        "results/",      % Desde directorio raíz
        "../results/",   % Desde test/ o scripts/
        "../../results/" % Desde subdirectorios anidados
    };
    
    results_dir = "";
    for i = 1:length(possible_dirs)
        if exist(possible_dirs{i}, 'dir')
            results_dir = possible_dirs{i};
            return;
        end
    end
    
    % Si no existe, intentar crear
    try
        mkdir("../results/");
        results_dir = "../results/";
    catch
        try
            mkdir("results/");
            results_dir = "results/";
        catch
            results_dir = "./"; % Último recurso
        end
    end
end

% --- Paso 1: Buscar archivo de datos ---
FILENAME = find_data_file();
if isempty(FILENAME)
    error('No se encontró el archivo de datos en ninguna ubicación.');
end

RESULTS_DIR = find_results_dir();
Z_THRESHOLD = 1e-6; 

disp(["Archivo de datos encontrado: ", FILENAME]);
disp(["Directorio de resultados: ", RESULTS_DIR]);
disp(["Script autosuficiente: Analizando archivo '", FILENAME, "'..."]);

% --- Paso 2: Calcular N automáticamente desde la cabecera ---
try
    fid = fopen(FILENAME, 'r');
    if (fid == -1)
        error(['No se pudo abrir el archivo: ', FILENAME]);
    end
    header_line = fgetl(fid);
    fclose(fid);
    
    columns = strsplit(strtrim(header_line(2:end))); 
    TOTAL_COLS = numel(columns);
    N_BODIES = (TOTAL_COLS - 4) / 4;
    
    if (mod(N_BODIES, 1) ~= 0 || N_BODIES <= 0)
        error('El número de cuerpos calculado desde la cabecera no es un entero positivo.');
    end
    
    disp(['Cuerpos detectados: N=', num2str(N_BODIES)]);
catch ME
    error(['No se pudo leer la cabecera del archivo para determinar N. Error: ', ME.message]);
end

% --- Paso 3: Cargar datos y detectar dimensionalidad ---
try
    data = dlmread(FILENAME, '', 1, 0); 
catch ME
    error(['No se pudo leer el archivo de datos. Error: ', ME.message]);
end

z_col_indices = 4:3:(1 + 3*N_BODIES);
z_data = data(:, z_col_indices);
z_range = max(z_data(:)) - min(z_data(:));
is_3d = z_range > Z_THRESHOLD;

if (is_3d)
    dim_string = '3D';
else
    dim_string = '2D';
end
disp(['Rango de Z detectado: ', num2str(z_range), '. El movimiento es ', dim_string, '.']);

% --- Paso 4: Graficar Trayectorias ---
figure('visible', 'off');
hold on;
colors = lines(N_BODIES); 
legends = {};

if (is_3d)
    disp('Generando gráfico de trayectorias 3D...');
    for i = 1:N_BODIES
        idx_x = 2 + 3*(i-1); idx_y = 3 + 3*(i-1); idx_z = 4 + 3*(i-1);
        plot3(data(:, idx_x), data(:, idx_y), data(:, idx_z), 'LineWidth', 2, 'Color', colors(i,:));
        legends{end+1} = ['Cuerpo ', num2str(i)];
    end
    xlabel('X'); ylabel('Y'); zlabel('Z');
    title(sprintf('Trayectorias (Vista 3D, N=%d Cuerpos)', N_BODIES));
    view(3);
    axis equal;
else
    disp('Generando gráfico de trayectorias 2D...');
    for i = 1:N_BODIES
        idx_x = 2 + 3*(i-1); idx_y = 3 + 3*(i-1);
        plot(data(:, idx_x), data(:, idx_y), 'LineWidth', 2, 'Color', colors(i,:));
        legends{end+1} = ['Cuerpo ', num2str(i)];
    end
    xlabel('X'); ylabel('Y');
    title(sprintf('Trayectorias (Vista 2D, N=%d Cuerpos)', N_BODIES));
    axis equal;
end
legend(legends);
grid on;
hold off;

output_path_traj = [RESULTS_DIR, sprintf('trayectorias_octave_%s_%d.png', dim_string, N_BODIES)];
print(output_path_traj, '-dpng');
disp(['Gráfica de trayectorias guardada en: ', output_path_traj]);

% --- Paso 5: Graficar Energías ---
col_k_idx = 1 + 4*N_BODIES + 1;
col_u_idx = col_k_idx + 1;
col_e_idx = col_k_idx + 2;

if col_e_idx <= size(data, 2)
    figure('visible', 'off');
    hold on;
    plot(data(:,1), data(:,col_k_idx), 'r', 'LineWidth', 2, 'DisplayName', 'Energía Cinética Total');
    plot(data(:,1), data(:,col_u_idx), 'b', 'LineWidth', 2, 'DisplayName', 'Energía Potencial Total');
    plot(data(:,1), data(:,col_e_idx), 'g', 'LineWidth', 2, 'DisplayName', 'Energía Total del Sistema');
    hold off;
    xlabel('Tiempo'); ylabel('Energía');
    title(sprintf('Energías del Sistema (N=%d Cuerpos)', N_BODIES));
    legend show;
    grid on;

    output_path_energy = [RESULTS_DIR, sprintf('energias_octave_%d.png', N_BODIES)];
    print(output_path_energy, '-dpng');
    disp(['Gráfica de energías guardada en: ', output_path_energy]);
else
    disp('No se encontraron columnas de energía en el archivo de datos.');
end

disp('Proceso de graficación completado.');
